### ISSUE TITLE
Problems with 6 dof model

### ISSUE THREAD

**Issue by jguarato** [petrobras/ross#1043]
Hi @raphaeltimbo,

As I mentioned in Issue #1034, I decided to conduct the same analyses performed in the ROSS tutorial part 2 [🔗](https://ross.readthedocs.io/en/latest/user_guide/tutorial_part_2.html) for the 6 dof model. The objective was to identify where errors could arise within the code. As some problems were indeed found, I opened this new issue to report them.

**Comment by jguarato** [comment]
Upon executing part 2 of the tutorial, I encountered the same error when invoking `run_static()` and `run_ucs()`:

```python
in Rotor.__init__(self, shaft_elements, disk_elements, bearing_elements, point_mass_elements, min_w, max_w, rated_w, tag)
    209 if len(tags_list) != len(set(tags_list)):
    210     raise ValueError("Tags should be unique.")
--> 212 self.number_dof = self._check_number_dof()
    214 ####################################################
    215 # Rotor summary
    216 ####################################################
    217 columns = [
    218     "type",
    219     "n",
   (...)
    237     "tag",
    238 ]
...
    562         "The number of degrees o freedom of all elements must be the same! There are BEARING elements with discrepant DoFs."
    563     )
    565 return int(number_dof)

Exception: The number of degrees o freedom of all elements must be the same! There are BEARING elements with discrepant DoFs.
```
which can be fixed by replacing the call of `BearingElement()`class with `elm.__class__()`. After fixing that, I compared the results obtained from the 4 dof model with the 6 dof model, and observed the following disparities:

- **Running modal analysis:**
![compare_1](https://github.com/petrobras/ross/assets/82293939/5f4fd8dd-2e16-4f8c-8a92-f4b97e3aaec1)
![compare_2](https://github.com/petrobras/ross/assets/82293939/e8479de8-e2f2-4ce7-98e9-193eec8ff4f8)

- **Plotting Campbell diagram:**
![compare_3](https://github.com/petrobras/ross/assets/82293939/f053f70b-8ba6-4fcd-aed9-7a8d75a4d2a7)

- **Plotting the deflected shape for the unbalance response:**
![compare_4](https://github.com/petrobras/ross/assets/82293939/886ed5a5-761b-495a-95d9-02009c151203)

- **Plotting the deflected shape for the unbalance response:**
![compare_5](https://github.com/petrobras/ross/assets/82293939/6418f53c-9e0d-496d-afc7-82711f7cfc1e)

**Comment by jguarato** [comment]
So, I started debugging the result classes in `results.py` to check if there was any coding that fixed the dof number to 4. Therefore, I needed to make some changes mainly in `Shape` and `ForcedResponseResults`. After that, I got the following results for the 6 dof model:
![newplot(1)](https://github.com/petrobras/ross/assets/82293939/c938f152-91d2-47f3-9eb1-d7d4b45d5e7f)
![newplot(2)](https://github.com/petrobras/ross/assets/82293939/bc9c2095-06c3-42f6-a3f0-7a4982e58538)
![newplot(3)](https://github.com/petrobras/ross/assets/82293939/87ef48d3-5ddd-488e-a2d9-887291b83442)
![newplot(4)](https://github.com/petrobras/ross/assets/82293939/23d902cc-4987-4148-bfd5-a141a48cbbaf)

**Comment by jguarato** [comment]
Investigating the code I realized that there was still a problem related to the 6 dof model. I started debugging the matrices of the axis elements by comparing them with the matrices of the 4 dof model. Consequently, I noticed that some signs in K matrix and M matrix did not coincide:
![K matrix](https://github.com/petrobras/ross/assets/82293939/18c488f2-049f-4946-bb83-3b3fc946cde3)

After correcting the corresponding signs, the following results were obtained:
![newplot](https://github.com/petrobras/ross/assets/82293939/23051cab-2239-4c2b-8a0e-efbab0c080fc)
![newplot(1)](https://github.com/petrobras/ross/assets/82293939/ea80f069-e5c5-4777-ac66-79455c523f32)
![newplot(2)](https://github.com/petrobras/ross/assets/82293939/6e0c4628-be7b-4647-9dbf-f457063c8a0c)
![newplot(3)](https://github.com/petrobras/ross/assets/82293939/c9c40d55-0ca1-4c8c-a7b3-c46569f64131)
![newplot(4)](https://github.com/petrobras/ross/assets/82293939/94a2d369-feea-42c9-b558-a0571c16c221)

**Comment by jguarato** [comment]
Finally, I would like to highlight two points:

- Apparently, most of the identified issues have been fixed. However, there is still an issue related to the UCS map. When running the UCS map for the 6 dof model passing different numbers of modes to the `run_ucs` function, the curves that appear on the map are not always smooth as for the 4 dof model. I will continue to investigate this.
![newplot](https://github.com/petrobras/ross/assets/82293939/ee7c1711-bc2a-4ecc-a695-84bc43d4a263)

- When I compared the M matrix of the shaft element of both models, I noticed another difference between them. In the 4 dof model, the shear effect is taken into account with the consideration of `phi` variable, however, in the 6 dof model, this effect is considered only in the K matrix but not in M matrix.

**Comment by jguarato** [comment]
Apparently, the problem in the UCS map has already been resolved with the correction applied in PR #931.

![ucs_6dof_24modes_fixed](https://github.com/petrobras/ross/assets/82293939/310a4fb0-0c9f-4321-b6cc-c80d67a7bcbd)

**Comment by jguarato** [comment]
n_links still have to be implemented in bearing element matrices for the 6 dof model. When running some analyses, I got errors related to singular matrices.

I was running the following example of the 4 dof model, but which I converted to the 6 dof model:

```python
n = 6

shaft_elem = [
    rs.ShaftElement(
        L=0.25,
        idl=0.0,
        odl=0.05,
        material=steel,
        shear_effects=True,
        rotary_inertia=True,
        gyroscopic=True,
    )
    for _ in range(n)
]

disk0 = rs.DiskElement.from_geometry(
    n=2, material=steel, width=0.07, i_d=0.05, o_d=0.28
)
disk1 = rs.DiskElement.from_geometry(
    n=4, material=steel, width=0.07, i_d=0.05, o_d=0.28
)
disks = [disk0, disk1]

stfx = 1e6
stfy = 0.8e6
bearing0 = rs.BearingElement(0, kxx=stfx, kyy=stfy, cxx=0, n_link=7)
bearing1 = rs.BearingElement(6, kxx=stfx, kyy=stfy, cxx=0)
bearing2 = rs.BearingElement(7, kxx=stfx, kyy=stfy, cxx=0)

bearings = [bearing0, bearing1, bearing2]

pm0 = rs.PointMass(n=7, m=30)
pointmass = [pm0]

rotor1 = rs.Rotor(shaft_elem, disks, bearings, pointmass)
```

First I tried the analyses: `run_static()`, and `run_modal()`, and in both cases the singular matrix error appeared.
